File: src/id0060.py
# Licensed under the MIT License.

# Prime Pair Sets

from math import inf, log10
from sympy import isprime, primerange
from time import time

MAX_SEARCH = 10000
    
def math_is_prime_pair(a: int, b: int) -> int: 
    return isprime(int(str(a) + str(b))) and isprime(int(str(b) + str(a)))

start = time()
primes = list(primerange(2, MAX_SEARCH))
min = inf

for i in range(len(primes)):
    a = primes[i]
    candidates = []
    
    for j in range(i + 1, len(primes)):
        b = primes[j]
        
        if not math_is_prime_pair(a, b): continue
        
        candidates.append(b)
    
    for j in range(len(candidates)):
        for k in range(j + 1, len(candidates)):
            b = candidates[j]
            c = candidates[k]
            
            if not math_is_prime_pair(b, c): continue
            
            for l in range(k + 1, len(candidates)):
                d = candidates[l]
                
                if not math_is_prime_pair(b, d): continue
                if not math_is_prime_pair(c, d): continue
                 
                for m in range(l + 1, len(candidates)):
                    e = candidates[m]
                    sum = a + b + c + d + e
                    
                    if sum >= min: continue
                    if not math_is_prime_pair(b, e): continue
                    if not math_is_prime_pair(c, e): continue
                    if not math_is_prime_pair(d, e): continue
                    
                    min = sum
    
print(f"0060{min:>64}    {time() - start:.6f}")

File: src/id0060.c
// Licensed under the MIT License.

// Prime Pair Sets

#include <limits.h>
#include <math.h>
#include "../lib/euler.h"
#include "../lib/sieve.h"
#define MAX_SEARCH 10000

static bool math_is_prime_pair(Sieve primes, int a, int b)
{
    long concat = math_concat(a, b);
    PrimalityTest primalityTest = miller_rabin_primality_test;

    if (sieve_test(primes, concat, primalityTest) != PRIMALITY_PRIME)
    {
        return false;
    }

    concat = math_concat(b, a);

    if (sieve_test(primes, concat, primalityTest) != PRIMALITY_PRIME)
    {
        return false;
    }

    return true;
}

int main(void)
{
    int min = INT_MAX;
    struct Sieve primes;
    struct List candidates;
    clock_t start = clock();

    euler_ok(sieve(&primes, MAX_SEARCH));
    euler_ok(list(&candidates, sizeof(long long), 0));

    long long* primesBegin = primes.primes.items;
    long long* primesEnd = primesBegin + primes.primes.count;

    for (long long* a = primesBegin; a < primesEnd; a++)
    {
        list_clear(&candidates);

        for (long long* b = a; b < primesEnd; b++)
        {
            if (math_is_prime_pair(&primes, *a, *b))
            {
                euler_ok(list_add(&candidates, b));
            }
        }

        long long* candidatesBegin = candidates.items;
        long long* candidatesEnd = candidatesBegin + candidates.count;

        for (long long* b = candidatesBegin; b < candidatesEnd; b++)
        {
            for (long long* c = b + 1; c < candidatesEnd; c++)
            {
                if (!math_is_prime_pair(&primes, *b, *c))
                {
                    continue;
                }

                for (long long* d = c + 1; d < candidatesEnd; d++)
                {
                    if (!math_is_prime_pair(&primes, *b, *d) ||
                        !math_is_prime_pair(&primes, *c, *d))
                    {
                        continue;
                    }

                    for (long long* e = d + 1; e < candidatesEnd; e++)
                    {
                        int sum = *a + *b + *c + *d + *e;

                        if (sum >= min ||
                            !math_is_prime_pair(&primes, *b, *e) ||
                            !math_is_prime_pair(&primes, *c, *e) ||
                            !math_is_prime_pair(&primes, *d, *e))
                        {
                            continue;
                        }

                        min = sum;
                    }
                }
            }
        }
    }

    finalize_sieve(&primes);
    finalize_list(&candidates);

    return euler_submit(60, min, start);
}
